node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,158 @@
 package cni
 
 import (
+	"context"
 	"encoding/base64"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
+	"sync"
 	"time"
 
 	"github.com/sirupsen/logrus"
-	"gopkg.in/fsnotify/fsnotify.v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 )
 
-var serviceaccountDirectory string = "/var/run/secrets/kubernetes.io/serviceaccount/"
+const defaultCNITokenValiditySeconds = 3600
+const minTokenRetryDuration = 5 * time.Second
+const defaultRefreshFraction = 3
+
 var kubeconfigPath string = "/host/etc/cni/net.d/calico-kubeconfig"
 
-func Run() {
-	// Log to stdout.  this prevents our logs from being interpreted as errors by, for example,
-	// fluentd's default configuration.
-	logrus.SetOutput(os.Stdout)
+type TokenRefresher struct {
+	tokenSupported bool
+	tokenOnce      *sync.Once
+
+	namespace          string
+	serviceAccountName string
 
-	// Create a watcher for file changes.
-	watcher, err := fsnotify.NewWatcher()
+	tokenChan chan TokenUpdate
+}
+
+type TokenUpdate struct {
+	Token          string
+	ExpirationTime time.Time
+	Error          error
+}
+
+func NewTokenRefresher(namespace string, serviceAccountName string) *TokenRefresher {
+	return &TokenRefresher{
+		tokenSupported:     false,
+		tokenOnce:          &sync.Once{},
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		tokenChan:          make(chan TokenUpdate),
+	}
+}
+
+func (t *TokenRefresher) UpdateToken() TokenUpdate {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.WithError(err).Error("Error generating kube config.")
+		return tokenUpdateError(err)
+	}
+	err = rest.LoadTLSFiles(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Error loading TLS files.")
+		return tokenUpdateError(err)
+	}
+	clientset, err := kubernetes.NewForConfig(cfg)
```

Comment:
Do we need to create a new clientset every iteration? 

IIUC, the Kubernetes clientset automatically knows how to refresh its credentials should they be rotated, so we don't need to do that here - or is there another reason to do this?

Reply:
If possible, passing a clientset to `NewTokenRefresher` would be ideal since we could use mock implementations for easier unit testing that way.